src/ledger-iframe-bridge.ts

```diff
@@ -0,0 +1,291 @@
+  iframe?: HTMLIFrameElement;
+
+  iframeLoaded = false;
+
+  bridgeUrl = '';
+
+  isDeviceConnected = false;
+
+  currentMessageId = 0;
+
+  messageCallbacks: Record<
+    number,
+    (response: IFrameMessageResponse<IFrameMessageAction>) => void
+  > = {};
+
+  delayedPromise?: {
+    resolve: (value: boolean) => void;
+    reject: (error: unknown) => void;
+    transportType: string;
+  };
+
+  async init(bridgeUrl: string) {
+    this.bridgeUrl = bridgeUrl;
+
+    this.#setupIframe();
+
+    window.addEventListener('message', this.#eventListener.bind(this));
+  }
+
+  async destroy() {
+    window.removeEventListener('message', this.#eventListener.bind(this));
+  }
+
+  async attemptMakeApp(): Promise<boolean> {
+    return new Promise((resolve, reject) => {
+      this.#sendMessage(
+        {
+          action: IFrameMessageAction.LedgerMakeApp,
+        },
+        (response) => {
+          if (response.success) {
+            resolve(true);
+          } else {
+            reject(response.error);
+          }
+        },
+      );
+    });
+  }
+
+  async updateTransportMethod(transportType: string): Promise<boolean> {
+    return new Promise((resolve, reject) => {
+      // If the iframe isn't loaded yet, let's store the desired transportType value and
+      // optimistically return a successful promise
+      if (!this.iframeLoaded) {
+        this.delayedPromise = {
+          resolve,
+          reject,
+          transportType,
+        };
+        return;
+      }
+
+      this.#sendMessage(
+        {
+          action: IFrameMessageAction.LedgerUpdateTransport,
+          params: { transportType },
+        },
+        ({ success }) => {
+          if (success) {
+            return resolve(true);
+          }
+          return reject(new Error('Ledger transport could not be updated'));
+        },
+      );
+    });
+  }
+
+  async getPublicKey(
+    params: GetPublicKeyParams,
+  ): Promise<GetPublicKeyResponse> {
+    return this.#deviceActionMessage(
+      IFrameMessageAction.LedgerUnlock,
+      params,
+    ) as Promise<GetPublicKeyResponse>;
+  }
+
+  async deviceSignTransaction(
+    params: LedgerSignTransactionParams,
+  ): Promise<LedgerSignTransactionResponse> {
+    return this.#deviceActionMessage(
+      IFrameMessageAction.LedgerSignTransaction,
+      params,
+    ) as Promise<LedgerSignTransactionResponse>;
+  }
+
+  async deviceSignMessage(
+    params: LedgerSignMessageParams,
+  ): Promise<LedgerSignMessageResponse> {
+    return this.#deviceActionMessage(
+      IFrameMessageAction.LedgerSignPersonalMessage,
+      params,
+    ) as Promise<LedgerSignMessageResponse>;
+  }
+
+  async deviceSignTypedData(
+    params: LedgerSignTypedDataParams,
+  ): Promise<LedgerSignTypedDataResponse> {
+    return this.#deviceActionMessage(
+      IFrameMessageAction.LedgerSignTypedData,
+      params,
+    ) as Promise<LedgerSignTypedDataResponse>;
+  }
+
+  async #deviceActionMessage(
+    action:
```

Comment:
The types you've specified here would allow an invalid combination of arguments to be provided. For instance, you could pass an `action` of `IFrameMessageAction.LedgerSignTransaction`, but `params` of `LedgerSignTypedDataParams`. What do you think about using overloads to specify the desired combinations instead? That might prevent the typecasting above.

Reply:
Thanks. I was thinking so hard on how to make typescript infer the return type with generics that I completely overlooked using overloads, which should work perfectly in this case.